### PR TITLE
add reapply_lost_speed to KinematicCharacterController

### PR DESCRIPTION
**Pull request by ntibi**
Adds `reapply_lost_speed: bool` to KinematicCharacterController

When set to true, the velocity lost by sliding will be reapplied to the corrected direction
so the character moves at the same speed wether it's sliding or not

### CODE REVIEW

File: src/control/character_controller.rs [hunk 1]
```diff
@@ -267,6 +272,17 @@ impl KinematicCharacterController {
...
+                    let diff =
+                        pre_slide_translation_remaining.norm() - translation_remaining.norm();
+                    if let Some((normalized_dir, _)) =
+                        // threshold has to be > subtract_hit hardcoded correction, if not, we get made up movement in the direction of the hit normal
+                        UnitVector::try_new_and_get(translation_remaining, 1.0e-4)
+                    {
+                        // reapply the lost speed (but in the corrected direction)
+                        translation_remaining += *normalized_dir * diff;
+                    }
+                }
```
Comment (Vrixyz) [inline src/control/character_controller.rs:285]: That condition should probably be in case there is no stairs ? otherwise we might end up intersecting with the stair. (and it might result in not going up the stair, or worse an infinite loop ?)

Comment (ntibi) [reply src/control/character_controller.rs]: indeed, but even with the reapply in the `if !self.handle_stairs`, reapplying the speed can be slightly flaky, especially in 3d

im turning the PR back into a draft and looking into it

**Comment by MeFisto94** [comment]
I guess kind of reworking the controller to not have the speed loss in the first place is not really possible? Would that result in loads of loop iterations of the slope sweeps?

I am asking because technically this behavior of slowing down/speeding up on slopes is not what one would expect from a character controller, especially contradicting the "kinematic" keyword. Same as the concept of "friction", it should usually always be infinite or zero.

For slopes specifically, the expectation (subjectively) is that one can climb them at full speed and descend them at full speed, unless the angle exceeds the maximum angle, _then_ slipping is involved in some way.

My problem currently also is that I can't control which slopes to climb as the effect of the slope angles is too high, while also causing considerate CPU load (the loop that does the offset by the normal nudge and then raytraces again).

So, TLDR: This sounds like it solves problems of an actual dynamic character controller, not a kinematic one and properly getting that kinematic behavior may involve going back to the drawing board

**Comment by sebcrozet** [comment]
Triaging. Thank you for this PR. I’m closing it since it’s been on draft for a while. If that remains a problem, please open an issue with a minimal scenes that would benefit from this.
